// @utoo/pack/cjs/utils/lockfile.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PersistentCacheLock = void 0;
exports.acquirePersistentCacheLock = acquirePersistentCacheLock;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const picocolors_1 = __importDefault(require("picocolors"));
const binding = __importStar(require("../binding"));
const RETRY_DELAY_MS = 10;
const MAX_RETRY_MS = 1000;
class PersistentCacheLock {
    constructor(nativeLockfile) {
        this.nativeLockfile = nativeLockfile;
        this.listener = () => this.unlockSync();
        process.on("exit", this.listener);
    }
    static tryAcquire(lockPath, content) {
        const nativeLockfile = binding.lockfileTryAcquireSync(lockPath, content);
        return nativeLockfile ? new PersistentCacheLock(nativeLockfile) : undefined;
    }
    static async acquireWithRetries(lockPath, processName) {
        const content = JSON.stringify({
            pid: process.pid,
            processName,
            startedAt: Date.now(),
        });
        const startMs = Date.now();
        let lockfile;
        while (Date.now() - startMs < MAX_RETRY_MS) {
            lockfile = PersistentCacheLock.tryAcquire(lockPath, content);
            if (lockfile) {
                return lockfile;
            }
            await new Promise((resolve) => setTimeout(resolve, RETRY_DELAY_MS));
        }
        throw new Error(formatLockError(lockPath, processName));
    }
    async unlock() {
        const lockfile = this.nativeLockfile;
        this.nativeLockfile = undefined;
        if (this.listener) {
            process.off("exit", this.listener);
            this.listener = undefined;
        }
        if (lockfile) {
            await binding.lockfileUnlock(lockfile);
        }
    }
    unlockSync() {
        const lockfile = this.nativeLockfile;
        this.nativeLockfile = undefined;
        if (this.listener) {
            process.off("exit", this.listener);
            this.listener = undefined;
        }
        if (lockfile) {
            binding.lockfileUnlockSync(lockfile);
        }
    }
}
exports.PersistentCacheLock = PersistentCacheLock;
async function acquirePersistentCacheLock(projectPath, processName, persistentCaching) {
    if (!persistentCaching) {
        return undefined;
    }
    const internalDir = path_1.default.join(path_1.default.resolve(projectPath), ".turbopack");
    fs_1.default.mkdirSync(internalDir, { recursive: true });
    return PersistentCacheLock.acquireWithRetries(path_1.default.join(internalDir, "lock"), processName);
}
function formatLockError(lockPath, processName) {
    let owner = "";
    try {
        const data = JSON.parse(fs_1.default.readFileSync(lockPath, "utf-8"));
        if (data.pid) {
            const killCommand = process.platform === "win32"
                ? `taskkill /PID ${data.pid} /F`
                : `kill ${data.pid}`;
            owner = `\nExisting process: ${data.processName || "unknown"} (PID ${data.pid}).\nStop it with ${picocolors_1.default.cyan(killCommand)} if it is stale.`;
        }
    }
    catch (_a) {
        // The lock holder might be a version that did not write metadata.
    }
    return `Unable to acquire ${processName} persistent cache lock at ${picocolors_1.default.cyan(lockPath)}. Another utoo pack process may be using the same .turbopack cache.${owner}`;
}
