// @utoo/pack/cjs/core/project.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurbopackInternalError = void 0;
exports.serializeConfig = serializeConfig;
exports.projectFactory = projectFactory;
const path_1 = require("path");
const util_1 = require("util");
const binding = __importStar(require("../binding"));
const common_1 = require("../utils/common");
const normalizePath_1 = require("../utils/normalizePath");
const loaderWorkerPool_1 = require("./loaderWorkerPool");
/**
 * An error caused by a bug in Turbopack, and not the user's code (e.g. a Rust panic). These should
 * be written to a log file and details should not be shown to the user.
 *
 * These are constructed in Turbopack by calling `throwTurbopackInternalError`.
 */
class TurbopackInternalError extends Error {
    constructor({ message, anonymizedLocation }) {
        super(message);
        this.name = "TurbopackInternalError";
        this.location = anonymizedLocation;
    }
}
exports.TurbopackInternalError = TurbopackInternalError;
/**
 * A helper used by the napi Rust entrypoints to construct and throw a `TurbopackInternalError`.
 */
function throwTurbopackInternalError(conversionError, opts) {
    if (conversionError != null) {
        throw new Error("NAPI type conversion error in throwTurbopackInternalError", { cause: conversionError });
    }
    throw new TurbopackInternalError(opts);
}
async function withErrorCause(fn) {
    var _a;
    try {
        return await fn();
    }
    catch (nativeError) {
        throw new TurbopackInternalError({
            message: (_a = nativeError === null || nativeError === void 0 ? void 0 : nativeError.message) !== null && _a !== void 0 ? _a : String(nativeError),
        });
    }
}
function normalizeEmotionConfig(emotion, isDev) {
    if (emotion === undefined || emotion === false) {
        return undefined;
    }
    const defaults = {
        sourcemap: isDev,
        autoLabel: isDev ? "always" : "never",
    };
    return emotion === true ? defaults : { ...defaults, ...emotion };
}
function normalizeStyles(styles, isDev) {
    if (!styles) {
        return styles;
    }
    return {
        ...styles,
        emotion: normalizeEmotionConfig(styles.emotion, isDev),
    };
}
// Align with next.js turbopack default optimizePackageImports config:
// https://nextjs.org/docs/app/api-reference/config/next-config-js/optimizePackageImports
const DEFAULT_OPTIMIZE_PACKAGE_IMPORTS = [
    "lucide-react",
    "date-fns",
    "lodash-es",
    "ramda",
    "antd",
    "react-bootstrap",
    "ahooks",
    "@ant-design/icons",
    "@headlessui/react",
    "@headlessui-float/react",
    "@heroicons/react/20/solid",
    "@heroicons/react/24/solid",
    "@heroicons/react/24/outline",
    "@visx/visx",
    "@tremor/react",
    "rxjs",
    "@mui/material",
    "@mui/icons-material",
    "recharts",
    "react-use",
    "effect",
    "@effect/schema",
    "@effect/platform",
    "@effect/platform-node",
    "@effect/platform-browser",
    "@effect/platform-bun",
    "@effect/sql",
    "@effect/sql-mssql",
    "@effect/sql-mysql2",
    "@effect/sql-pg",
    "@effect/sql-sqlite-node",
    "@effect/sql-sqlite-bun",
    "@effect/sql-sqlite-wasm",
    "@effect/sql-sqlite-react-native",
    "@effect/rpc",
    "@effect/rpc-http",
    "@effect/typeclass",
    "@effect/experimental",
    "@effect/opentelemetry",
    "@material-ui/core",
    "@material-ui/icons",
    "@tabler/icons-react",
    "mui-core",
    "react-icons/ai",
    "react-icons/bi",
    "react-icons/bs",
    "react-icons/cg",
    "react-icons/ci",
    "react-icons/di",
    "react-icons/fa",
    "react-icons/fa6",
    "react-icons/fc",
    "react-icons/fi",
    "react-icons/gi",
    "react-icons/go",
    "react-icons/gr",
    "react-icons/hi",
    "react-icons/hi2",
    "react-icons/im",
    "react-icons/io",
    "react-icons/io5",
    "react-icons/lia",
    "react-icons/lib",
    "react-icons/lu",
    "react-icons/md",
    "react-icons/pi",
    "react-icons/ri",
    "react-icons/rx",
    "react-icons/si",
    "react-icons/sl",
    "react-icons/tb",
    "react-icons/tfi",
    "react-icons/ti",
    "react-icons/vsc",
    "react-icons/wi",
];
function mergePackageImports(packageImports) {
    const defaultPackageImports = process.env.UTOO_DISABLE_DEFAULT_PACKAGE_IMPORTS === "1"
        ? []
        : DEFAULT_OPTIMIZE_PACKAGE_IMPORTS;
    return [...new Set([...(packageImports !== null && packageImports !== void 0 ? packageImports : []), ...defaultPackageImports])];
}
function normalizeOptimization(optimization) {
    return {
        ...(optimization !== null && optimization !== void 0 ? optimization : {}),
        packageImports: mergePackageImports(optimization === null || optimization === void 0 ? void 0 : optimization.packageImports),
    };
}
async function serializeConfig(config, isDev) {
    const configSerializable = {
        ...config,
        styles: normalizeStyles(config.styles, isDev),
        optimization: normalizeOptimization(config.optimization),
    };
    if (configSerializable.entry) {
        configSerializable.entry = configSerializable.entry.map((entry) => {
            const { html, ...rest } = entry;
            return rest;
        });
    }
    {
        const { modularizeImports } = configSerializable.optimization;
        if (modularizeImports) {
            configSerializable.optimization.modularizeImports = Object.fromEntries(Object.entries(modularizeImports).map(([mod, config]) => [
                mod,
                {
                    ...config,
                    transform: typeof config.transform === "string"
                        ? config.transform
                        : Object.entries(config.transform).map(([key, value]) => [
                            key,
                            value,
                        ]),
                },
            ]));
        }
    }
    if (configSerializable.module && configSerializable.module.rules) {
        configSerializable.module.rules = serializeModuleRules(configSerializable.module.rules);
    }
    return JSON.stringify(configSerializable, null, 2);
}
// converts regexes to a `RegexComponents` object so that it can be JSON-serialized when passed to
// Turbopack
function serializeRuleCondition(cond) {
    function regexComponents(regex) {
        return {
            source: regex.source,
            flags: regex.flags,
        };
    }
    if (typeof cond === "string") {
        return cond;
    }
    else if ("all" in cond) {
        return { ...cond, all: cond.all.map(serializeRuleCondition) };
    }
    else if ("any" in cond) {
        return { ...cond, any: cond.any.map(serializeRuleCondition) };
    }
    else if ("not" in cond) {
        return { ...cond, not: serializeRuleCondition(cond.not) };
    }
    else {
        return {
            ...cond,
            path: cond.path == null
                ? undefined
                : cond.path instanceof RegExp
                    ? {
                        type: "regex",
                        value: regexComponents(cond.path),
                    }
                    : { type: "glob", value: cond.path },
            content: cond.content && regexComponents(cond.content),
            query: cond.query == null
                ? undefined
                : cond.query instanceof RegExp
                    ? {
                        type: "regex",
                        value: regexComponents(cond.query),
                    }
                    : { type: "constant", value: cond.query },
            contentType: cond.contentType == null
                ? undefined
                : cond.contentType instanceof RegExp
                    ? {
                        type: "regex",
                        value: regexComponents(cond.contentType),
                    }
                    : { type: "glob", value: cond.contentType },
        };
    }
}
// Note: Returns an updated `turbopackRules` with serialized conditions. Does not mutate in-place.
function serializeModuleRules(turbopackRules) {
    const serializedRules = {};
    for (const [glob, rule] of Object.entries(turbopackRules)) {
        if (Array.isArray(rule)) {
            serializedRules[glob] = rule.map((item) => {
                if (typeof item !== "string" &&
                    ("loaders" in item || "type" in item || "condition" in item)) {
                    return serializeConfigItem(item, glob);
                }
                else {
                    checkLoaderItem(item, glob);
                    return item;
                }
            });
        }
        else {
            serializedRules[glob] = serializeConfigItem(rule, glob);
        }
    }
    return serializedRules;
    function serializeConfigItem(rule, glob) {
        if (!rule)
            return rule;
        if (rule.loaders) {
            for (const item of rule.loaders) {
                checkLoaderItem(item, glob);
            }
        }
        let serializedRule = rule;
        if (rule.condition != null) {
            serializedRule = {
                ...rule,
                condition: serializeRuleCondition(rule.condition),
            };
        }
        return serializedRule;
    }
    function checkLoaderItem(loaderItem, glob) {
        if (typeof loaderItem !== "string" &&
            !(0, util_1.isDeepStrictEqual)(loaderItem, JSON.parse(JSON.stringify(loaderItem)))) {
            throw new Error(`loader ${loaderItem.loader} for match "${glob}" does not have serializable options. ` +
                "Ensure that options passed are plain JavaScript objects and values.");
        }
    }
}
async function rustifyPartialProjectOptions(options) {
    return {
        ...options,
        rootPath: normalizePathOption(options.rootPath),
        projectPath: normalizePathOption(options.projectPath),
        packPath: normalizePathOption(options.packPath),
        config: options.config && (await serializeConfig(options.config, !!options.dev)),
        processEnv: options.processEnv && (0, common_1.rustifyEnv)(options.processEnv),
    };
}
function normalizePathOption(pathLike) {
    if (pathLike === undefined)
        return undefined;
    return (0, normalizePath_1.normalizePath)(pathLike);
}
async function rustifyProjectOptions(options) {
    var _a;
    return {
        ...options,
        rootPath: (0, normalizePath_1.normalizePath)(options.rootPath),
        projectPath: (0, normalizePath_1.normalizePath)(options.projectPath),
        packPath: (0, normalizePath_1.normalizePath)(options.packPath),
        config: await serializeConfig(options.config, options.dev),
        processEnv: (0, common_1.rustifyEnv)((_a = options.processEnv) !== null && _a !== void 0 ? _a : {}),
    };
}
function projectFactory() {
    const cancel = new (class Cancel extends Error {
    })();
    function subscribe(useBuffer, nativeFunction) {
        // A buffer of produced items. This will only contain values if the
        // consumer is slower than the producer.
        let buffer = [];
        // A deferred value waiting for the next produced item. This will only
        // exist if the consumer is faster than the producer.
        let waiting;
        let canceled = false;
        // The native function will call this every time it emits a new result. We
        // either need to notify a waiting consumer, or buffer the new result until
        // the consumer catches up.
        function emitResult(err, value) {
            if (waiting) {
                let { resolve, reject } = waiting;
                waiting = undefined;
                if (err)
                    reject(err);
                else
                    resolve(value);
            }
            else {
                const item = { err, value };
                if (useBuffer)
                    buffer.push(item);
                else
                    buffer[0] = item;
            }
        }
        async function* createIterator() {
            const task = await withErrorCause(() => nativeFunction(emitResult));
            try {
                while (!canceled) {
                    if (buffer.length > 0) {
                        const item = buffer.shift();
                        if (item.err)
                            throw item.err;
                        yield item.value;
                    }
                    else {
                        // eslint-disable-next-line no-loop-func
                        yield new Promise((resolve, reject) => {
                            waiting = { resolve, reject };
                        });
                    }
                }
            }
            catch (e) {
                if (e === cancel)
                    return;
                if (e instanceof Error) {
                    throw new TurbopackInternalError({ message: e.message });
                }
                throw e;
            }
            finally {
                if (task) {
                    binding.rootTaskDispose(task);
                }
            }
        }
        const iterator = createIterator();
        iterator.return = async () => {
            canceled = true;
            if (waiting)
                waiting.reject(cancel);
            return { value: undefined, done: true };
        };
        return iterator;
    }
    class ProjectImpl {
        constructor(nativeProject) {
            this._nativeProject = nativeProject;
            if (typeof binding.registerWorkerScheduler === "function") {
                (0, loaderWorkerPool_1.runLoaderWorkerPool)(binding, (0, path_1.resolve)((0, common_1.getPackPath)(), "./binding.js"));
            }
        }
        async update(options) {
            await withErrorCause(async () => binding.projectUpdate(this._nativeProject, await rustifyPartialProjectOptions(options)));
        }
        async writeAllEntrypointsToDisk() {
            return await withErrorCause(async () => {
                const napiEndpoints = (await binding.projectWriteAllEntrypointsToDisk(this._nativeProject));
                return napiEntrypointsToRawEntrypoints(napiEndpoints);
            });
        }
        entrypointsSubscribe() {
            const subscription = subscribe(false, async (callback) => binding.projectEntrypointsSubscribe(this._nativeProject, callback));
            return (async function* () {
                for await (const entrypoints of subscription) {
                    yield napiEntrypointsToRawEntrypoints(entrypoints);
                }
            })();
        }
        hmrEvents(identifier) {
            return subscribe(true, async (callback) => binding.projectHmrEvents(this._nativeProject, identifier, callback));
        }
        hmrIdentifiersSubscribe() {
            return subscribe(false, async (callback) => binding.projectHmrIdentifiersSubscribe(this._nativeProject, callback));
        }
        traceSource(stackFrame, currentDirectoryFileUrl) {
            return binding.projectTraceSource(this._nativeProject, stackFrame, currentDirectoryFileUrl);
        }
        getSourceForAsset(filePath) {
            return binding.projectGetSourceForAsset(this._nativeProject, filePath);
        }
        getSourceMap(filePath) {
            return binding.projectGetSourceMap(this._nativeProject, filePath);
        }
        getSourceMapSync(filePath) {
            return binding.projectGetSourceMapSync(this._nativeProject, filePath);
        }
        updateInfoSubscribe(aggregationMs) {
            return subscribe(true, async (callback) => binding.projectUpdateInfoSubscribe(this._nativeProject, aggregationMs, callback));
        }
        shutdown() {
            return binding.projectShutdown(this._nativeProject);
        }
        onExit() {
            return binding.projectOnExit(this._nativeProject);
        }
    }
    class EndpointImpl {
        constructor(nativeEndpoint) {
            this._nativeEndpoint = nativeEndpoint;
        }
        async writeToDisk() {
            return await withErrorCause(() => binding.endpointWriteToDisk(this._nativeEndpoint));
        }
        async clientChanged() {
            const clientSubscription = subscribe(false, async (callback) => binding.endpointClientChangedSubscribe(await this._nativeEndpoint, callback));
            await clientSubscription.next();
            return clientSubscription;
        }
        async serverChanged(includeIssues) {
            const serverSubscription = subscribe(false, async (callback) => binding.endpointServerChangedSubscribe(await this._nativeEndpoint, includeIssues, callback));
            await serverSubscription.next();
            return serverSubscription;
        }
    }
    function napiEntrypointsToRawEntrypoints(entrypoints) {
        return {
            apps: (entrypoints.apps || []).map((e) => new EndpointImpl(e)),
            libraries: (entrypoints.libraries || []).map((e) => new EndpointImpl(e)),
            issues: entrypoints.issues,
            diagnostics: entrypoints.diagnostics,
        };
    }
    return async function createProject(options, turboEngineOptions) {
        return new ProjectImpl(await binding.projectNew(await rustifyProjectOptions(options), turboEngineOptions || {}, {
            throwTurbopackInternalError,
        }));
    };
}
