// @utoo/pack/esm/core/hmr.js
import { formatIssue, HMR_ACTIONS_SENT_TO_BROWSER, } from "@utoo/pack-shared";
import { nanoid } from "nanoid";
import { WebSocketServer } from "ws";
import { HtmlPlugin } from "../plugins/HtmlPlugin.js";
import { cleanOutput, getOutputPath } from "../utils/cleanOutput.js";
import { debounce, getPackPath, processIssues } from "../utils/common.js";
import { getInitialAssetsFromStats } from "../utils/getInitialAssets.js";
import { processHtmlEntry } from "../utils/htmlEntry.js";
import { acquirePersistentCacheLock } from "../utils/lockfile.js";
import { normalizePath } from "../utils/normalizePath.js";
import { useWorkerThreads } from "../utils/runtimePluginStratety.js";
import { validateEntryPaths } from "../utils/validateEntry.js";
import { projectFactory } from "./project.js";
const wsServer = new WebSocketServer({ noServer: true });
const sessionId = Math.floor(Number.MAX_SAFE_INTEGER * Math.random());
// Re-export HMR types from pack-shared for backward compatibility
export { HMR_ACTIONS_SENT_TO_BROWSER, } from "@utoo/pack-shared";
export const FAST_REFRESH_RUNTIME_RELOAD = "Fast Refresh had to perform a full reload due to a runtime error.";
function hasBlockingIssues(issues) {
    for (const issueMap of issues.values()) {
        for (const issue of issueMap.values()) {
            if (issue.severity !== "warning") {
                return true;
            }
        }
    }
    return false;
}
function hasBlockingResultIssues(result) {
    return result.issues.some((issue) => issue.severity === "error" || issue.severity === "fatal");
}
function addIssuesToErrors(errors, issues) {
    for (const issueMap of issues.values()) {
        for (const [key, issue] of issueMap) {
            if (issue.severity === "warning") {
                continue;
            }
            errors.set(key, {
                message: formatIssue(issue, false),
            });
        }
    }
}
function getCompilationErrors(issues) {
    const errors = new Map();
    addIssuesToErrors(errors, issues);
    return [...errors.values()];
}
function getClientIssueKey(id) {
    return `client:${id}`;
}
export async function createHotReloader(bundleOptions, projectPath, rootPath) {
    var _a, _b, _c, _d, _e, _f, _g;
    const resolvedProjectPath = projectPath || process.cwd();
    const resolvedRootPath = rootPath || projectPath || process.cwd();
    processHtmlEntry(bundleOptions.config, resolvedProjectPath);
    validateEntryPaths(bundleOptions.config, resolvedProjectPath);
    await cleanOutput(bundleOptions.config, resolvedProjectPath);
    const createProject = projectFactory();
    const persistentCaching = (_a = bundleOptions.config.persistentCaching) !== null && _a !== void 0 ? _a : false;
    const persistentCacheLock = await acquirePersistentCacheLock(resolvedProjectPath, "utoo pack dev", persistentCaching);
    const htmlConfigs = [
        ...(Array.isArray(bundleOptions.config.html)
            ? bundleOptions.config.html
            : bundleOptions.config.html
                ? [bundleOptions.config.html]
                : []),
        ...bundleOptions.config.entry
            .filter((e) => !!e.html)
            .map((e) => e.html),
    ];
    const shouldCreateWebpackStats = Boolean(process.env.ANALYZE) ||
        bundleOptions.config.stats ||
        htmlConfigs.length > 0;
    let project;
    try {
        project = await createProject({
            processEnv: (_b = bundleOptions.processEnv) !== null && _b !== void 0 ? _b : {},
            watch: {
                enable: true,
            },
            dev: true,
            buildId: bundleOptions.buildId || nanoid(),
            tracing: (_c = bundleOptions.tracing) !== null && _c !== void 0 ? _c : true,
            config: {
                ...bundleOptions.config,
                mode: "development",
                stats: shouldCreateWebpackStats,
                optimization: {
                    ...bundleOptions.config.optimization,
                    minify: false,
                    moduleIds: "named",
                },
                persistentCaching: (_e = (_d = bundleOptions === null || bundleOptions === void 0 ? void 0 : bundleOptions.config) === null || _d === void 0 ? void 0 : _d.persistentCaching) !== null && _e !== void 0 ? _e : true,
                pluginRuntimeStrategy: (_g = (_f = bundleOptions === null || bundleOptions === void 0 ? void 0 : bundleOptions.config) === null || _f === void 0 ? void 0 : _f.pluginRuntimeStrategy) !== null && _g !== void 0 ? _g : (useWorkerThreads() ? "workerThreads" : "childProcesses"),
            },
            projectPath: normalizePath(resolvedProjectPath),
            rootPath: resolvedRootPath,
            packPath: getPackPath(),
        }, {
            persistentCaching,
        });
    }
    catch (error) {
        persistentCacheLock === null || persistentCacheLock === void 0 ? void 0 : persistentCacheLock.unlockSync();
        throw error;
    }
    const entrypointsSubscription = project.entrypointsSubscribe();
    let currentEntriesHandlingResolve;
    let currentEntriesHandling = new Promise((resolve) => (currentEntriesHandlingResolve = resolve));
    let hmrEventHappened = false;
    let hmrHash = 0;
    const clients = new Set();
    const clientStates = new WeakMap();
    const currentEntryIssues = new Map();
    const backgroundWatchSubscriptions = new Set();
    let currentWatchedEntrypoints = [];
    let backgroundWatchersStarted = false;
    let backgroundWatchGeneration = 0;
    const backgroundEndpointWriteTasks = new Map();
    let backgroundProjectWriteTask;
    let closed = false;
    let closePromise;
    function sendToClient(client, payload) {
        client.send(JSON.stringify(payload));
    }
    function sendEnqueuedMessages() {
        if (hasBlockingIssues(currentEntryIssues)) {
            return;
        }
        for (const client of clients) {
            const state = clientStates.get(client);
            if (!state) {
                continue;
            }
            if (hasBlockingIssues(state.clientIssues)) {
                return;
            }
            for (const payload of state.hmrPayloads.values()) {
                sendToClient(client, payload);
            }
            state.hmrPayloads.clear();
            if (state.turbopackUpdates.length > 0) {
                sendToClient(client, {
                    action: HMR_ACTIONS_SENT_TO_BROWSER.TURBOPACK_MESSAGE,
                    data: state.turbopackUpdates,
                });
                state.turbopackUpdates.length = 0;
            }
        }
    }
    const sendEnqueuedMessagesDebounce = debounce(sendEnqueuedMessages, 2);
    function sendTurbopackMessage(payload) {
        var _a;
        payload.diagnostics = [];
        payload.issues = [];
        for (const client of clients) {
            (_a = clientStates.get(client)) === null || _a === void 0 ? void 0 : _a.turbopackUpdates.push(payload);
        }
        hmrEventHappened = true;
        sendEnqueuedMessagesDebounce();
    }
    async function regenerateHtml() {
        var _a;
        if (htmlConfigs.length === 0) {
            return;
        }
        const outputDir = getOutputPath(bundleOptions.config, resolvedProjectPath);
        const publicPath = (_a = bundleOptions.config.output) === null || _a === void 0 ? void 0 : _a.publicPath;
        const assets = getInitialAssetsFromStats(outputDir);
        for (const config of htmlConfigs) {
            const plugin = new HtmlPlugin(config);
            await plugin.generate(outputDir, assets, publicPath);
        }
    }
    async function writeAllEntrypointsToDisk() {
        const result = await project.writeAllEntrypointsToDisk();
        processIssues(result, true, true);
        await regenerateHtml();
    }
    async function writeEntrypointToDisk(entrypoint) {
        const result = await entrypoint.writeToDisk();
        processIssues(result, true, true);
        await regenerateHtml();
    }
    async function writeOutputToDisk(entrypoint) {
        if (shouldCreateWebpackStats) {
            await writeAllEntrypointsToDisk();
        }
        else {
            await writeEntrypointToDisk(entrypoint);
        }
    }
    async function disposeBackgroundWatchSubscriptions() {
        const subscriptions = [...backgroundWatchSubscriptions];
        backgroundWatchSubscriptions.clear();
        currentEntryIssues.clear();
        backgroundEndpointWriteTasks.clear();
        backgroundProjectWriteTask = undefined;
        await Promise.all(subscriptions.map((subscription) => { var _a; return (_a = subscription.return) === null || _a === void 0 ? void 0 : _a.call(subscription); }));
    }
    function scheduleOutputWrite(entrypoint, generation) {
        var _a;
        if (!backgroundWatchersStarted ||
            closed ||
            generation !== backgroundWatchGeneration) {
            return;
        }
        const previousTask = shouldCreateWebpackStats
            ? (backgroundProjectWriteTask !== null && backgroundProjectWriteTask !== void 0 ? backgroundProjectWriteTask : Promise.resolve())
            : ((_a = backgroundEndpointWriteTasks.get(entrypoint)) !== null && _a !== void 0 ? _a : Promise.resolve());
        const task = previousTask
            .catch(() => { })
            .then(async () => {
            await currentEntriesHandling;
            if (closed || generation !== backgroundWatchGeneration) {
                return;
            }
            await writeOutputToDisk(entrypoint);
            hmrEventHappened = true;
        })
            .finally(() => {
            if (shouldCreateWebpackStats) {
                if (backgroundProjectWriteTask === task) {
                    backgroundProjectWriteTask = undefined;
                }
            }
            else if (backgroundEndpointWriteTasks.get(entrypoint) === task) {
                backgroundEndpointWriteTasks.delete(entrypoint);
            }
        });
        if (shouldCreateWebpackStats) {
            backgroundProjectWriteTask = task;
        }
        else {
            backgroundEndpointWriteTasks.set(entrypoint, task);
        }
    }
    async function refreshBackgroundWatchers() {
        const generation = ++backgroundWatchGeneration;
        await disposeBackgroundWatchSubscriptions();
        if (!backgroundWatchersStarted || closed) {
            return;
        }
        await Promise.all(currentWatchedEntrypoints.map(async (entrypoint) => {
            var _a, _b;
            const [clientChanges, serverChanges] = await Promise.all([
                entrypoint.clientChanged(),
                entrypoint.serverChanged(true),
            ]);
            if (closed || generation !== backgroundWatchGeneration) {
                await Promise.all([
                    (_a = clientChanges.return) === null || _a === void 0 ? void 0 : _a.call(clientChanges),
                    (_b = serverChanges.return) === null || _b === void 0 ? void 0 : _b.call(serverChanges),
                ]);
                return;
            }
            backgroundWatchSubscriptions.add(clientChanges);
            backgroundWatchSubscriptions.add(serverChanges);
            const watchChanges = async (subscription, issueKey) => {
                try {
                    for await (const data of subscription) {
                        if (closed || generation !== backgroundWatchGeneration) {
                            return;
                        }
                        processIssues(currentEntryIssues, issueKey, data, false, true);
                        if (hasBlockingResultIssues(data)) {
                            hmrEventHappened = true;
                            sendEnqueuedMessagesDebounce();
                            continue;
                        }
                        scheduleOutputWrite(entrypoint, generation);
                    }
                }
                catch (error) {
                    if (!closed && generation === backgroundWatchGeneration) {
                        console.error(error);
                        process.exit(1);
                    }
                }
                finally {
                    backgroundWatchSubscriptions.delete(subscription);
                    currentEntryIssues.delete(issueKey);
                }
            };
            const entrypointIndex = currentWatchedEntrypoints.indexOf(entrypoint);
            void watchChanges(clientChanges, `entrypoint:${entrypointIndex}:client`);
            void watchChanges(serverChanges, `entrypoint:${entrypointIndex}:server`);
        }));
    }
    async function subscribeToHmrEvents(client, id) {
        const state = clientStates.get(client);
        if (!state || state.subscriptions.has(id)) {
            return;
        }
        const subscription = project.hmrEvents(id);
        state.subscriptions.set(id, subscription);
        const issueKey = getClientIssueKey(id);
        // The subscription will always emit once, which is the initial
        // computation. This is not a change, so swallow it.
        try {
            await subscription.next();
            for await (const data of subscription) {
                processIssues(state.clientIssues, issueKey, data, false, true);
                if (data.type !== "issues") {
                    sendTurbopackMessage(data);
                }
            }
        }
        catch (e) {
            // The client might be using an HMR session from a previous server, tell them
            // to fully reload the page to resolve the issue. We can't use
            // `hotReloader.send` since that would force every connected client to
            // reload, only this client is out of date.
            const reloadAction = {
                action: HMR_ACTIONS_SENT_TO_BROWSER.RELOAD,
                data: `error in HMR event subscription for ${id}: ${e}`,
            };
            sendToClient(client, reloadAction);
            client.close();
            return;
        }
    }
    function unsubscribeFromHmrEvents(client, id) {
        const state = clientStates.get(client);
        if (!state) {
            return;
        }
        const subscription = state.subscriptions.get(id);
        subscription === null || subscription === void 0 ? void 0 : subscription.return();
        state.clientIssues.delete(getClientIssueKey(id));
    }
    async function handleEntrypointsSubscription() {
        var _a, _b;
        for await (const entrypoints of entrypointsSubscription) {
            if (!currentEntriesHandlingResolve) {
                currentEntriesHandling = new Promise(
                // eslint-disable-next-line no-loop-func
                (resolve) => (currentEntriesHandlingResolve = resolve));
            }
            currentWatchedEntrypoints = [
                ...((_a = entrypoints.apps) !== null && _a !== void 0 ? _a : []),
                ...((_b = entrypoints.libraries) !== null && _b !== void 0 ? _b : []),
            ];
            if (shouldCreateWebpackStats) {
                await writeAllEntrypointsToDisk();
            }
            else {
                await Promise.all(currentWatchedEntrypoints.map((entrypoint) => writeEntrypointToDisk(entrypoint)));
            }
            if (backgroundWatchersStarted) {
                await refreshBackgroundWatchers();
            }
            currentEntriesHandlingResolve();
            currentEntriesHandlingResolve = undefined;
        }
    }
    const hotReloader = {
        turbopackProject: project,
        serverStats: null,
        onHMR(req, socket, head, onUpgrade) {
            wsServer.handleUpgrade(req, socket, head, (client) => {
                onUpgrade === null || onUpgrade === void 0 ? void 0 : onUpgrade(client);
                const subscriptions = new Map();
                clients.add(client);
                clientStates.set(client, {
                    hmrPayloads: new Map(),
                    turbopackUpdates: [],
                    subscriptions,
                    clientIssues: new Map(),
                });
                client.on("close", () => {
                    var _a;
                    // Remove active subscriptions
                    for (const subscription of subscriptions.values()) {
                        (_a = subscription.return) === null || _a === void 0 ? void 0 : _a.call(subscription);
                    }
                    clientStates.delete(client);
                    clients.delete(client);
                });
                client.addEventListener("message", ({ data }) => {
                    const parsedData = JSON.parse(typeof data !== "string" ? data.toString() : data);
                    // messages
                    switch (parsedData.event) {
                        case "client-error": // { errorCount, clientId }
                        case "client-warning": // { warningCount, clientId }
                        case "client-success": // { clientId }
                        case "client-full-reload": // { stackTrace, hadRuntimeError }
                            const { hadRuntimeError, dependencyChain } = parsedData;
                            if (hadRuntimeError) {
                                console.warn(FAST_REFRESH_RUNTIME_RELOAD);
                            }
                            if (Array.isArray(dependencyChain) &&
                                typeof dependencyChain[0] === "string") {
                                const cleanedModulePath = dependencyChain[0]
                                    .replace(/^\[project\]/, ".")
                                    .replace(/ \[.*\] \(.*\)$/, "");
                                console.warn(`Fast Refresh had to perform a full reload when ${cleanedModulePath} changed.`);
                            }
                            break;
                        default:
                            // Might be a Turbopack message...
                            if (!parsedData.type) {
                                throw new Error(`unrecognized HMR message "${data}"`);
                            }
                    }
                    // Turbopack messages
                    switch (parsedData.type) {
                        case "turbopack-subscribe":
                            subscribeToHmrEvents(client, parsedData.path);
                            break;
                        case "turbopack-unsubscribe":
                            unsubscribeFromHmrEvents(client, parsedData.path);
                            break;
                        default:
                            if (!parsedData.event) {
                                throw new Error(`unrecognized Turbopack HMR message "${data}"`);
                            }
                    }
                });
                const turbopackConnected = {
                    action: HMR_ACTIONS_SENT_TO_BROWSER.TURBOPACK_CONNECTED,
                    data: { sessionId },
                };
                sendToClient(client, turbopackConnected);
                const errors = getCompilationErrors(currentEntryIssues);
                (async function () {
                    const sync = {
                        action: HMR_ACTIONS_SENT_TO_BROWSER.SYNC,
                        errors,
                        warnings: [],
                        hash: "",
                    };
                    sendToClient(client, sync);
                })();
            });
        },
        registerClient(ws) {
            const subscriptions = new Map();
            clients.add(ws);
            clientStates.set(ws, {
                hmrPayloads: new Map(),
                turbopackUpdates: [],
                subscriptions,
                clientIssues: new Map(),
            });
            const turbopackConnected = {
                action: HMR_ACTIONS_SENT_TO_BROWSER.TURBOPACK_CONNECTED,
                data: { sessionId },
            };
            sendToClient(ws, turbopackConnected);
            const errors = getCompilationErrors(currentEntryIssues);
            const sync = {
                action: HMR_ACTIONS_SENT_TO_BROWSER.SYNC,
                errors,
                warnings: [],
                hash: "",
            };
            sendToClient(ws, sync);
        },
        unregisterClient(ws) {
            var _a;
            const state = clientStates.get(ws);
            if (state) {
                for (const subscription of state.subscriptions.values()) {
                    (_a = subscription.return) === null || _a === void 0 ? void 0 : _a.call(subscription);
                }
            }
            clientStates.delete(ws);
            clients.delete(ws);
        },
        handleClientMessage(ws, data) {
            const parsedData = JSON.parse(data);
            switch (parsedData.event) {
                case "client-error":
                case "client-warning":
                case "client-success":
                case "client-full-reload": {
                    const { hadRuntimeError, dependencyChain } = parsedData;
                    if (hadRuntimeError) {
                        console.warn(FAST_REFRESH_RUNTIME_RELOAD);
                    }
                    if (Array.isArray(dependencyChain) &&
                        typeof dependencyChain[0] === "string") {
                        const cleanedModulePath = dependencyChain[0]
                            .replace(/^\[project\]/, ".")
                            .replace(/ \[.*\] \(.*\)$/, "");
                        console.warn(`Fast Refresh had to perform a full reload when ${cleanedModulePath} changed.`);
                    }
                    break;
                }
                default:
                    if (!parsedData.type) {
                        throw new Error(`unrecognized HMR message "${data}"`);
                    }
            }
            switch (parsedData.type) {
                case "turbopack-subscribe":
                    subscribeToHmrEvents(ws, parsedData.path);
                    break;
                case "turbopack-unsubscribe":
                    unsubscribeFromHmrEvents(ws, parsedData.path);
                    break;
                default:
                    if (!parsedData.event) {
                        throw new Error(`unrecognized Turbopack HMR message "${data}"`);
                    }
            }
        },
        send(action) {
            const payload = JSON.stringify(action);
            for (const client of clients) {
                client.send(payload);
            }
        },
        setHmrServerError(_error) {
            // Not implemented yet.
        },
        clearHmrServerError() {
            // Not implemented yet.
        },
        async start() {
            if (backgroundWatchersStarted) {
                return;
            }
            backgroundWatchersStarted = true;
            await refreshBackgroundWatchers();
        },
        async buildFallbackError() {
            // Not implemented yet.
        },
        close() {
            closed = true;
            void disposeBackgroundWatchSubscriptions();
            closePromise !== null && closePromise !== void 0 ? closePromise : (closePromise = project
                .onExit()
                .catch((err) => {
                console.error(err);
            })
                .finally(() => {
                persistentCacheLock === null || persistentCacheLock === void 0 ? void 0 : persistentCacheLock.unlockSync();
            }));
            for (const wsClient of clients) {
                wsClient.close();
            }
            clients.clear();
        },
    };
    handleEntrypointsSubscription().catch((err) => {
        console.error(err);
        process.exit(1);
    });
    // Write empty manifests
    await currentEntriesHandling;
    async function handleProjectUpdates() {
        for await (const updateMessage of project.updateInfoSubscribe(30)) {
            switch (updateMessage.updateType) {
                case "start": {
                    hotReloader.send({ action: HMR_ACTIONS_SENT_TO_BROWSER.BUILDING });
                    break;
                }
                case "end": {
                    sendEnqueuedMessages();
                    const errors = new Map();
                    addIssuesToErrors(errors, currentEntryIssues);
                    for (const client of clients) {
                        const state = clientStates.get(client);
                        if (!state) {
                            continue;
                        }
                        const clientErrors = new Map(errors);
                        addIssuesToErrors(clientErrors, state.clientIssues);
                        sendToClient(client, {
                            action: HMR_ACTIONS_SENT_TO_BROWSER.BUILT,
                            hash: String(++hmrHash),
                            errors: [...clientErrors.values()],
                            warnings: [],
                        });
                    }
                    if (hmrEventHappened) {
                        const time = updateMessage.value.duration;
                        const timeMessage = time > 2000 ? `${Math.round(time / 100) / 10}s` : `${time}ms`;
                        console.log(`Compiled in ${timeMessage}`);
                        hmrEventHappened = false;
                    }
                    break;
                }
                default:
            }
        }
    }
    handleProjectUpdates().catch((err) => {
        console.error(err);
        process.exit(1);
    });
    return hotReloader;
}
