// @umijs/bundler-utoopack/dist/config.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/config.ts
var config_exports = {};
__export(config_exports, {
  getDevUtooPackConfig: () => getDevUtooPackConfig,
  getProdUtooPackConfig: () => getProdUtooPackConfig,
  getSSRUtooPackConfig: () => getSSRUtooPackConfig,
  mergeExtraPostcssPlugins: () => mergeExtraPostcssPlugins,
  normalizeUtoopackPath: () => normalizeUtoopackPath
});
module.exports = __toCommonJS(config_exports);
var import_bundler_webpack = require("@umijs/bundler-webpack");
var import_utils = require("@umijs/utils");
var import_pack = require("@utoo/pack");
var import_fs = __toESM(require("fs"));
var import_path = require("path");
var DEFAULT_STATIC_PATH_PREFIX = "static/";
function getAssetModuleFilename(staticPathPrefix) {
  const prefix = staticPathPrefix !== void 0 ? staticPathPrefix : DEFAULT_STATIC_PATH_PREFIX;
  return `${prefix}[name].[contenthash:8]`;
}
function getUtoopackDefine(opts) {
  const define = Object.fromEntries(
    Object.entries(opts.config.define || {}).map(([key, value]) => {
      return [key, normalizeUtoopackDefineValue(value)];
    })
  );
  if (process.env.SOCKET_SERVER) {
    define["process.env.SOCKET_SERVER"] = JSON.stringify(
      process.env.SOCKET_SERVER
    );
  }
  return define;
}
function normalizeUtoopackDefineValue(value) {
  if (typeof value === "string") {
    return JSON.stringify(value);
  }
  return value;
}
function getModularizeImports(extraBabelPlugins) {
  return extraBabelPlugins.filter((p) => /^import$|babel-plugin-import/.test(p[0])).reduce(
    (acc, [_, v]) => {
      const {
        libraryName,
        libraryDirectory,
        style,
        camel2DashComponentName,
        transformToDefaultImport,
        ...rest
      } = v;
      if (Object.keys(rest).length > 0) {
        throw new Error(
          `babel-plugin-import options ${Object.keys(
            rest
          )} is not supported in utoopack bundler`
        );
      }
      if (typeof style === "function") {
        throw new Error(
          `babel-plugin-import style function is not supported in utoopack bundler`
        );
      }
      let transformRule = "{{ kebabCase member }}";
      if (camel2DashComponentName === false) {
        transformRule = "{{ member }}";
      }
      const skipDefaultConversion = typeof transformToDefaultImport === "undefined" ? false : !Boolean(transformToDefaultImport);
      acc[libraryName] = {
        transform: `${libraryName}/${libraryDirectory}/${transformRule}`,
        preventFullImport: false,
        skipDefaultConversion,
        style: typeof style === "boolean" ? "style" : style
      };
      return acc;
    },
    {}
  );
}
function getBabelPluginName(plugin) {
  const name = Array.isArray(plugin) ? plugin[0] : plugin;
  return typeof name === "string" ? name : "";
}
function isModularizeImportPlugin(plugin) {
  return /^import$|babel-plugin-import/.test(getBabelPluginName(plugin));
}
function isEmotionBabelPlugin(plugin) {
  const name = getBabelPluginName(plugin);
  return name === "@emotion" || name.endsWith("@emotion/babel-plugin");
}
function getExtraBabelPlugins(opts) {
  return [
    ...opts.beforeBabelPlugins || [],
    ...opts.extraBabelPlugins || [],
    ...opts.config.extraBabelPlugins || []
  ].filter(Boolean).filter(isSerializableBabelItem);
}
function getExtraBabelPresets(opts) {
  return [
    ...opts.beforeBabelPresets || [],
    ...opts.extraBabelPresets || [],
    ...opts.config.extraBabelPresets || []
  ].filter(Boolean).filter(isSerializableBabelItem);
}
function dropUndefinedValues(value) {
  if (Array.isArray(value)) {
    return value.map(dropUndefinedValues);
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).filter(([, item]) => item !== void 0).map(([key, item]) => [key, dropUndefinedValues(item)])
    );
  }
  return value;
}
function isSerializableBabelItem(item) {
  const normalized = dropUndefinedValues(item);
  try {
    return import_utils.lodash.isEqual(normalized, JSON.parse(JSON.stringify(normalized)));
  } catch (e) {
    return false;
  }
}
function getExtraBabelModuleRules(opts) {
  var _a;
  if (((_a = opts.config.utoopack) == null ? void 0 : _a.babelLoader) !== true) {
    return {};
  }
  const plugins = getExtraBabelPlugins(opts).filter((plugin) => {
    return !isModularizeImportPlugin(plugin) && !isEmotionBabelPlugin(plugin);
  });
  const extraPresets = getExtraBabelPresets(opts);
  if (!plugins.length && !extraPresets.length) {
    return {};
  }
  const presets = [opts.babelPreset, ...extraPresets].filter(Boolean);
  const rule = {
    condition: {
      all: [
        { not: "foreign" },
        {
          not: {
            path: /[\\/]src[\\/]\.umi(?:-[^\\/]*)?[\\/]/
          }
        }
      ]
    },
    loaders: [
      {
        loader: require.resolve("@umijs/bundler-webpack/compiled/babel-loader"),
        options: dropUndefinedValues({
          sourceType: "unambiguous",
          babelrc: false,
          configFile: false,
          cacheDirectory: false,
          browserslistConfigFile: false,
          targets: opts.config.targets,
          customize: opts.config.babelLoaderCustomize,
          presets,
          plugins
        })
      }
    ],
    as: "*.js"
  };
  return {
    module: {
      rules: Object.fromEntries(
        ["js", "mjs", "cjs", "jsx", "ts", "tsx"].map((ext) => [
          `**/src/**/*.${ext}`,
          rule
        ])
      )
    }
  };
}
function normalizeUtoopackPath(path) {
  return path.replace(/\\/g, "/");
}
function normalizeUtoopackEntry(entry) {
  if (typeof entry === "string") {
    return normalizeUtoopackPath(entry);
  }
  if (Array.isArray(entry)) {
    return entry.map(normalizeUtoopackEntry);
  }
  if (entry && typeof entry === "object") {
    return Object.fromEntries(
      Object.entries(entry).map(([key, value]) => [
        key,
        normalizeUtoopackEntry(value)
      ])
    );
  }
  return entry;
}
function normalizeUtoopackOpts(opts) {
  return {
    ...opts,
    cwd: normalizeUtoopackPath(opts.cwd),
    rootDir: normalizeUtoopackPath(opts.rootDir),
    entry: normalizeUtoopackEntry(opts.entry)
  };
}
function getNormalizedAlias(alias, rootDir) {
  const newAlias = Object.fromEntries(
    Object.entries(alias || {}).map(([key, value]) => [
      normalizeUtoopackPath(key),
      normalizeUtoopackPath(value)
    ])
  );
  const normalizedRootDir = normalizeUtoopackPath(rootDir);
  for (const [key, value] of Object.entries(newAlias)) {
    if (key.endsWith("/*") || value.endsWith("/*") || key.endsWith("/") || value.endsWith("/") || key.endsWith("$")) {
      continue;
    }
    const ext = (0, import_path.extname)(value);
    if (ext) {
      let isDirectory = false;
      const candidates = [
        .../* @__PURE__ */ new Set([value, (0, import_path.resolve)(normalizedRootDir, value)])
      ];
      for (const candidate of candidates) {
        try {
          const stat = import_fs.default.statSync(candidate);
          if (stat.isDirectory()) {
            isDirectory = true;
            break;
          }
        } catch (e) {
        }
      }
      if (!isDirectory) continue;
    }
    newAlias[`${key}/*`] = `${value}/*`;
  }
  newAlias[`${normalizedRootDir}/*`] = `${normalizedRootDir}/*`;
  return newAlias;
}
function normalizeExternalValue(v) {
  if (Array.isArray(v)) {
    const [url, ...members] = v;
    const scriptPrefix = /^script\s+/.exec(url);
    const script = scriptPrefix ? url.slice(scriptPrefix[0].length) : url;
    if (scriptPrefix) {
      return {
        // ['antd', 'Button'] => `antd.Button`
        root: members.join("."),
        type: "script",
        // `script https://example.com/lib/script.js` => `https://example.com/lib/script.js`
        script
      };
    } else if (url === "promise" && typeof members[0] === "string") {
      return `promise ${members[0]}`;
    } else {
      return {
        root: members.join("."),
        script
      };
    }
  } else if (typeof v === "string") {
    return v.replace(/^window(\s+|\.)/, "");
  }
}
function toWildcardExternalConfig(value) {
  const normalized = normalizeExternalValue(value);
  if (typeof normalized === "string") {
    const externalType = /^(commonjs|esm|promise)\s+/.exec(normalized);
    if (externalType) {
      return {
        root: normalized.slice(externalType[0].length),
        type: externalType[1]
      };
    }
    return {
      root: normalized
    };
  }
}
function getExternalSubPathGlob(externalKey) {
  if (!externalKey.includes("*")) return;
  const parts = externalKey.split("/");
  const packageName = externalKey.startsWith("@") ? parts.slice(0, 2).join("/") : parts[0];
  const subPathGlob = externalKey.slice(packageName.length);
  if (!packageName || !subPathGlob.startsWith("/")) return;
  return {
    packageName,
    subPathGlob
  };
}
function escapeRegexChar(char) {
  return /[|\\{}()[\]^$+?.]/.test(char) ? `\\${char}` : char;
}
function globToSubPathRegex(glob) {
  let regex = "^";
  for (const char of glob) {
    if (char === "*") {
      regex += ".+";
    } else if (char === "/") {
      regex += "\\/";
    } else {
      regex += escapeRegexChar(char);
    }
  }
  return `/${regex}$/`;
}
function addWildcardSubPath(config, subPathGlob) {
  var _a;
  return {
    ...config,
    subPath: {
      ...config.subPath || {},
      rules: [
        ...((_a = config.subPath) == null ? void 0 : _a.rules) || [],
        {
          regex: globToSubPathRegex(subPathGlob),
          target: ""
        }
      ]
    }
  };
}
function getNormalizedExternals(externals) {
  const ret = Object.entries(externals || {}).reduce(
    (memo, [k, v]) => {
      if (!getExternalSubPathGlob(k)) {
        const normalized = normalizeExternalValue(v);
        if (normalized !== void 0) {
          memo[k] = normalized;
        }
      }
      return memo;
    },
    {}
  );
  Object.entries(externals || {}).forEach(([k, v]) => {
    const externalSubPathGlob = getExternalSubPathGlob(k);
    if (!externalSubPathGlob) return;
    const normalized = toWildcardExternalConfig(v);
    if (!normalized) return;
    ret[externalSubPathGlob.packageName] = addWildcardSubPath(
      ret[externalSubPathGlob.packageName] && typeof ret[externalSubPathGlob.packageName] === "object" ? ret[externalSubPathGlob.packageName] : normalized,
      externalSubPathGlob.subPathGlob
    );
  });
  return ret;
}
function getSvgModuleRules(opts) {
  const { svgr, svgo = {} } = opts;
  if (!svgr) {
    return {};
  }
  return {
    module: {
      rules: {
        "*.svg": {
          loaders: [
            {
              loader: require.resolve("@umijs/bundler-webpack/dist/loader/svgr"),
              options: {
                svgoConfig: {
                  plugins: [
                    {
                      name: "preset-default",
                      params: {
                        overrides: {
                          removeTitle: false
                        }
                      }
                    },
                    "prefixIds"
                  ],
                  ...typeof svgo === "object" ? svgo : {}
                },
                ...svgr,
                svgo: !!svgo
              }
            },
            {
              loader: require.resolve("@umijs/bundler-webpack/compiled/url-loader")
            }
          ],
          as: "*.js"
        }
      }
    }
  };
}
function appendPostcssPlugin(postcssConfig, plugin) {
  const [pluginName, pluginOptions = {}] = plugin;
  if (postcssConfig == null) {
    return {
      plugins: {
        [pluginName]: pluginOptions
      }
    };
  }
  if (!import_utils.lodash.isPlainObject(postcssConfig)) {
    throw new Error(
      `Utoopack styles.postcss must be an object when extraPostCSSPlugins is used.`
    );
  }
  const existingPlugins = postcssConfig.plugins;
  if (existingPlugins != null && !import_utils.lodash.isPlainObject(existingPlugins)) {
    throw new Error(
      `Utoopack styles.postcss.plugins must be an object when extraPostCSSPlugins is used.`
    );
  }
  return {
    ...postcssConfig,
    plugins: {
      ...existingPlugins || {},
      [pluginName]: pluginOptions
    }
  };
}
function normalizeExtraPostcssPlugin(plugin) {
  if (typeof plugin === "string") {
    return [[plugin, {}]];
  }
  if (Array.isArray(plugin) && typeof plugin[0] === "string" && plugin.length <= 2) {
    return [[plugin[0], plugin[1] ?? {}]];
  }
  if (import_utils.lodash.isPlainObject(plugin)) {
    return Object.entries(plugin);
  }
  throw new Error(
    `Utoopack only supports JSON-serializable extraPostCSSPlugins entries. Please use plugin names or [pluginName, options] tuples instead.`
  );
}
function mergeExtraPostcssPlugins(postcssConfig, extraPlugins = []) {
  return extraPlugins.reduce((memo, plugin) => {
    return normalizeExtraPostcssPlugin(plugin).reduce(
      (ret, normalizedPlugin) => {
        return appendPostcssPlugin(ret, normalizedPlugin);
      },
      memo
    );
  }, postcssConfig);
}
function getUserUtoopackConfig(utoopackConfig = {}) {
  return import_utils.lodash.omit(utoopackConfig, ["babelLoader", "root"]);
}
function getDefaultPersistentCaching() {
  return process.platform !== "win32";
}
async function getProdUtooPackConfig(rawOpts) {
  var _a;
  const opts = normalizeUtoopackOpts(rawOpts);
  const webpackConfig = await (0, import_bundler_webpack.getConfig)({
    cwd: opts.cwd,
    rootDir: opts.rootDir,
    env: "production",
    entry: opts.entry,
    userConfig: opts.config,
    analyze: process.env.ANALYZE,
    babelPreset: opts.babelPreset,
    extraBabelPlugins: [
      ...opts.beforeBabelPlugins || [],
      ...opts.extraBabelPlugins || []
    ],
    extraBabelPresets: [
      ...opts.beforeBabelPresets || [],
      ...opts.extraBabelPresets || []
    ],
    extraBabelIncludes: opts.config.extraBabelIncludes,
    chainWebpack: opts.chainWebpack,
    modifyWebpackConfig: opts.modifyWebpackConfig,
    staticPathPrefix: opts.staticPathPrefix,
    pkg: opts.pkg,
    disableCopy: opts.disableCopy
  });
  let utooBundlerOpts = (0, import_pack.compatOptionsFromWebpack)({
    ...import_utils.lodash.omit(webpackConfig, ["target", "module", "externals"]),
    webpackMode: true
  });
  const extraBabelPlugins = [
    ...opts.extraBabelPlugins || [],
    ...opts.config.extraBabelPlugins || []
  ];
  const modularizeImports = getModularizeImports(extraBabelPlugins);
  const emotion = extraBabelPlugins.some((p) => {
    return p === "@emotion" || p === "@emotion/babel-plugin";
  });
  const define = getUtoopackDefine(opts);
  const {
    publicPath,
    runtimePublicPath,
    externals: userExternals,
    copy = [],
    svgr,
    svgo = {},
    inlineLimit,
    mdx
  } = opts.config;
  const userUtoopackConfig = getUserUtoopackConfig(opts.config.utoopack);
  utooBundlerOpts = {
    ...utooBundlerOpts,
    tracing: false,
    config: import_utils.lodash.merge(
      import_utils.lodash.omit(utooBundlerOpts.config, ["define"]),
      {
        output: {
          clean: opts.clean,
          publicPath: runtimePublicPath ? "runtime" : publicPath || "/",
          assetModuleFilename: getAssetModuleFilename(opts.staticPathPrefix),
          ...opts.disableCopy ? { copy: [] } : { copy: ["public"].concat(copy) }
        },
        optimization: {
          modularizeImports,
          concatenateModules: true
        },
        resolve: {
          alias: getNormalizedAlias(
            (_a = utooBundlerOpts.config.resolve) == null ? void 0 : _a.alias,
            opts.rootDir
          )
        },
        styles: {
          less: {
            modifyVars: opts.config.theme,
            javascriptEnabled: true,
            ...opts.config.lessLoader
          },
          // postcss: normalizedPostcssConfig,
          sass: opts.config.sassLoader ?? void 0,
          emotion
        },
        define,
        nodePolyfill: true,
        mdx: !!mdx,
        externals: getNormalizedExternals(userExternals)
      },
      getExtraBabelModuleRules(opts),
      getSvgModuleRules({ svgr, svgo, inlineLimit }),
      userUtoopackConfig
    )
  };
  return utooBundlerOpts;
}
async function getSSRUtooPackConfig(rawOpts) {
  const opts = normalizeUtoopackOpts(rawOpts);
  const utooBundlerOpts = await getProdUtooPackConfig({
    ...opts,
    clean: false,
    disableCopy: true
  });
  const entry = Object.entries(opts.entry)[0];
  const entryName = (entry == null ? void 0 : entry[0]) || "umi.server";
  const entryPath = entry == null ? void 0 : entry[1];
  const filename = opts.useHash ? "[name].[contenthash:8].js" : (0, import_path.basename)(opts.serverBuildPath);
  utooBundlerOpts.config = {
    ...utooBundlerOpts.config,
    entry: [
      {
        name: entryName,
        import: entryPath,
        library: {}
      }
    ],
    output: {
      ...utooBundlerOpts.config.output,
      path: (0, import_path.dirname)(opts.serverBuildPath),
      filename,
      chunkFilename: filename,
      clean: true,
      copy: [],
      publicPath: "/"
    },
    target: "node",
    sourceMaps: false,
    stats: true,
    nodePolyfill: false,
    optimization: {
      ...utooBundlerOpts.config.optimization,
      minify: false
    }
  };
  return utooBundlerOpts;
}
async function getDevUtooPackConfig(rawOpts) {
  var _a;
  const opts = normalizeUtoopackOpts(rawOpts);
  let webpackConfig = await (0, import_bundler_webpack.getConfig)({
    cwd: opts.cwd,
    rootDir: opts.rootDir,
    env: "development",
    entry: opts.entry,
    userConfig: opts.config,
    babelPreset: opts.babelPreset,
    extraBabelPlugins: [
      ...opts.beforeBabelPlugins || [],
      ...opts.extraBabelPlugins || []
    ],
    extraBabelPresets: [
      ...opts.beforeBabelPresets || [],
      ...opts.extraBabelPresets || []
    ],
    extraBabelIncludes: opts.config.extraBabelIncludes,
    chainWebpack: opts.chainWebpack,
    modifyWebpackConfig: opts.modifyWebpackConfig,
    staticPathPrefix: opts.staticPathPrefix,
    // TO avoild bundler webpack add extra entry.
    hmr: false,
    analyze: process.env.ANALYZE
  });
  let utooBundlerOpts = (0, import_pack.compatOptionsFromWebpack)({
    ...import_utils.lodash.omit(webpackConfig, ["target", "module", "externals"]),
    webpackMode: true
  });
  const extraBabelPlugins = [
    ...opts.extraBabelPlugins || [],
    ...opts.config.extraBabelPlugins || []
  ];
  const modularizeImports = getModularizeImports(extraBabelPlugins);
  const emotion = extraBabelPlugins.some((p) => {
    return p === "@emotion" || p === "@emotion/babel-plugin";
  });
  const define = getUtoopackDefine(opts);
  const {
    publicPath,
    runtimePublicPath,
    externals: userExternals,
    copy = [],
    svgr,
    svgo = {},
    inlineLimit,
    mdx
  } = opts.config;
  const userUtoopackConfig = getUserUtoopackConfig(opts.config.utoopack);
  utooBundlerOpts = {
    ...utooBundlerOpts,
    ...process.env.SOCKET_SERVER ? {
      processEnv: {
        ...utooBundlerOpts.processEnv || {},
        "process.env.SOCKET_SERVER": process.env.SOCKET_SERVER
      }
    } : {},
    config: import_utils.lodash.merge(
      import_utils.lodash.omit(utooBundlerOpts.config, ["define"]),
      {
        output: {
          clean: opts.clean === void 0 ? true : opts.clean,
          publicPath: runtimePublicPath ? "runtime" : publicPath || "/",
          assetModuleFilename: getAssetModuleFilename(opts.staticPathPrefix),
          ...opts.disableCopy ? { copy: [] } : { copy: ["public"].concat(copy) }
        },
        resolve: {
          alias: getNormalizedAlias(
            (_a = utooBundlerOpts.config.resolve) == null ? void 0 : _a.alias,
            opts.rootDir
          )
        },
        optimization: {
          modularizeImports,
          // Disable the options under dev for utoopack
          // otherwise it will affect react-refresh
          removeUnusedExports: false,
          removeUnusedImports: false
        },
        styles: {
          less: {
            modifyVars: opts.config.theme,
            javascriptEnabled: true,
            ...opts.config.lessLoader
          },
          // postcss: normalizedPostcssPlugin,
          sass: opts.config.sassLoader ?? void 0,
          emotion
        },
        define,
        stats: true,
        // Windows persistent cache restore is currently unstable in utoopack dev.
        persistentCaching: getDefaultPersistentCaching(),
        nodePolyfill: true,
        mdx: !!mdx,
        externals: getNormalizedExternals(userExternals),
        ...opts.config.clickToComponent ? {
          // clickToComponent relies on source filename metadata in dev.
          react: {
            absoluteSourceFilename: true
          }
        } : {}
      },
      getExtraBabelModuleRules(opts),
      getSvgModuleRules({ svgr, svgo, inlineLimit }),
      userUtoopackConfig
    ),
    watch: {
      enable: true
    },
    dev: true,
    tracing: false
  };
  return utooBundlerOpts;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getDevUtooPackConfig,
  getProdUtooPackConfig,
  getSSRUtooPackConfig,
  mergeExtraPostcssPlugins,
  normalizeUtoopackPath
});
