// @utoo/pack/esm/utils/lockfile.js
import fs from "fs";
import path from "path";
import pc from "picocolors";
import * as binding from "../binding.js";
const RETRY_DELAY_MS = 10;
const MAX_RETRY_MS = 1000;
export class PersistentCacheLock {
    constructor(nativeLockfile) {
        this.nativeLockfile = nativeLockfile;
        this.listener = () => this.unlockSync();
        process.on("exit", this.listener);
    }
    static tryAcquire(lockPath, content) {
        const nativeLockfile = binding.lockfileTryAcquireSync(lockPath, content);
        return nativeLockfile ? new PersistentCacheLock(nativeLockfile) : undefined;
    }
    static async acquireWithRetries(lockPath, processName) {
        const content = JSON.stringify({
            pid: process.pid,
            processName,
            startedAt: Date.now(),
        });
        const startMs = Date.now();
        let lockfile;
        while (Date.now() - startMs < MAX_RETRY_MS) {
            lockfile = PersistentCacheLock.tryAcquire(lockPath, content);
            if (lockfile) {
                return lockfile;
            }
            await new Promise((resolve) => setTimeout(resolve, RETRY_DELAY_MS));
        }
        throw new Error(formatLockError(lockPath, processName));
    }
    async unlock() {
        const lockfile = this.nativeLockfile;
        this.nativeLockfile = undefined;
        if (this.listener) {
            process.off("exit", this.listener);
            this.listener = undefined;
        }
        if (lockfile) {
            await binding.lockfileUnlock(lockfile);
        }
    }
    unlockSync() {
        const lockfile = this.nativeLockfile;
        this.nativeLockfile = undefined;
        if (this.listener) {
            process.off("exit", this.listener);
            this.listener = undefined;
        }
        if (lockfile) {
            binding.lockfileUnlockSync(lockfile);
        }
    }
}
export async function acquirePersistentCacheLock(projectPath, processName, persistentCaching) {
    if (!persistentCaching) {
        return undefined;
    }
    const internalDir = path.join(path.resolve(projectPath), ".turbopack");
    fs.mkdirSync(internalDir, { recursive: true });
    return PersistentCacheLock.acquireWithRetries(path.join(internalDir, "lock"), processName);
}
function formatLockError(lockPath, processName) {
    let owner = "";
    try {
        const data = JSON.parse(fs.readFileSync(lockPath, "utf-8"));
        if (data.pid) {
            const killCommand = process.platform === "win32"
                ? `taskkill /PID ${data.pid} /F`
                : `kill ${data.pid}`;
            owner = `\nExisting process: ${data.processName || "unknown"} (PID ${data.pid}).\nStop it with ${pc.cyan(killCommand)} if it is stale.`;
        }
    }
    catch (_a) {
        // The lock holder might be a version that did not write metadata.
    }
    return `Unable to acquire ${processName} persistent cache lock at ${pc.cyan(lockPath)}. Another utoo pack process may be using the same .turbopack cache.${owner}`;
}
