// @utoo/pack/esm/commands/build.js
import { handleIssues } from "@utoo/pack-shared";
import { spawn } from "child_process";
import fs from "fs";
import { nanoid } from "nanoid";
import path from "path";
import { resolveBundleOptions } from "../config/webpackCompat.js";
import { projectFactory } from "../core/project.js";
import { HtmlPlugin } from "../plugins/HtmlPlugin.js";
import { cleanOutput, getOutputPath } from "../utils/cleanOutput.js";
import { blockStdout, getPackPath } from "../utils/common.js";
import { findRootDir } from "../utils/findRoot.js";
import { getInitialAssetsFromStats } from "../utils/getInitialAssets.js";
import { processHtmlEntry } from "../utils/htmlEntry.js";
import { acquirePersistentCacheLock } from "../utils/lockfile.js";
import { normalizePath } from "../utils/normalizePath.js";
import { useWorkerThreads } from "../utils/runtimePluginStratety.js";
import { validateEntryPaths } from "../utils/validateEntry.js";
import { xcodeProfilingReady } from "../utils/xcodeProfile.js";
export function build(options, projectPath, rootPath) {
    const bundleOptions = resolveBundleOptions(options, projectPath, rootPath);
    if (!rootPath) {
        // help user to find the rootDir automatically.
        rootPath = findRootDir(projectPath || process.cwd());
    }
    return buildInternal(bundleOptions, projectPath, rootPath);
}
async function buildInternal(bundleOptions, projectPath, rootPath) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    blockStdout();
    if (process.env.XCODE_PROFILE) {
        await xcodeProfilingReady();
    }
    const resolvedProjectPath = projectPath || process.cwd();
    const resolvedRootPath = rootPath || projectPath || process.cwd();
    const persistentCaching = (_a = bundleOptions.config.persistentCaching) !== null && _a !== void 0 ? _a : false;
    processHtmlEntry(bundleOptions.config, resolvedProjectPath);
    validateEntryPaths(bundleOptions.config, resolvedProjectPath);
    await cleanOutput(bundleOptions.config, resolvedProjectPath);
    const createProject = projectFactory();
    const persistentCacheLock = await acquirePersistentCacheLock(resolvedProjectPath, "utoo pack build", persistentCaching);
    let project;
    try {
        project = await createProject({
            processEnv: (_b = bundleOptions.processEnv) !== null && _b !== void 0 ? _b : {},
            watch: {
                enable: false,
            },
            dev: (_c = bundleOptions.dev) !== null && _c !== void 0 ? _c : false,
            buildId: bundleOptions.buildId || nanoid(),
            tracing: (_d = bundleOptions.tracing) !== null && _d !== void 0 ? _d : true,
            config: {
                ...bundleOptions.config,
                stats: Boolean(process.env.ANALYZE) ||
                    bundleOptions.config.stats ||
                    bundleOptions.config.entry.some((e) => !!e.html),
                pluginRuntimeStrategy: (_f = (_e = bundleOptions === null || bundleOptions === void 0 ? void 0 : bundleOptions.config) === null || _e === void 0 ? void 0 : _e.pluginRuntimeStrategy) !== null && _f !== void 0 ? _f : (useWorkerThreads() ? "workerThreads" : "childProcesses"),
            },
            projectPath: normalizePath(resolvedProjectPath),
            rootPath: resolvedRootPath,
            packPath: getPackPath(),
        }, {
            persistentCaching,
            // Build mode is a short-lived, one-shot compilation, so avoid paying
            // dependency graph bookkeeping cost unless the persistent cache needs it.
            dependencyTracking: persistentCaching,
            isShortSession: true,
        });
        const entrypoints = await project.writeAllEntrypointsToDisk();
        handleIssues(entrypoints.issues);
        const htmlConfigs = [
            ...(Array.isArray(bundleOptions.config.html)
                ? bundleOptions.config.html
                : bundleOptions.config.html
                    ? [bundleOptions.config.html]
                    : []),
            ...bundleOptions.config.entry
                .filter((e) => !!e.html)
                .map((e) => e.html),
        ];
        if (htmlConfigs.length > 0) {
            const assets = { js: [], css: [] };
            const outputDir = getOutputPath(bundleOptions.config, resolvedProjectPath);
            if (assets.js.length === 0 && assets.css.length === 0) {
                const discovered = getInitialAssetsFromStats(outputDir);
                assets.js.push(...discovered.js);
                assets.css.push(...discovered.css);
            }
            const publicPath = (_g = bundleOptions.config.output) === null || _g === void 0 ? void 0 : _g.publicPath;
            for (const config of htmlConfigs) {
                const plugin = new HtmlPlugin(config);
                await plugin.generate(outputDir, assets, publicPath);
            }
        }
        if (process.env.ANALYZE) {
            await analyzeBundle(((_h = bundleOptions.config.output) === null || _h === void 0 ? void 0 : _h.path) || "dist");
        }
    }
    finally {
        await (project === null || project === void 0 ? void 0 : project.shutdown());
        persistentCacheLock === null || persistentCacheLock === void 0 ? void 0 : persistentCacheLock.unlockSync();
    }
    // TODO: Maybe run tasks in worker is a better way, see
    // https://github.com/vercel/next.js/blob/512d8283054407ab92b2583ecce3b253c3be7b85/packages/next/src/lib/worker.ts
}
async function analyzeBundle(outputPath) {
    const statsPath = path.join(outputPath, "stats.json");
    if (!fs.existsSync(statsPath)) {
        console.warn(`Stats file not found at ${statsPath}. Make sure to enable stats in your configuration.`);
        return;
    }
    return new Promise((resolve, reject) => {
        const analyzer = spawn("npx", ["webpack-bundle-analyzer", statsPath], {
            stdio: "inherit",
            shell: true,
        });
        analyzer.on("error", (error) => {
            reject(new Error(`Failed to start bundle analyzer: ${error.message}`));
        });
        analyzer.on("close", () => {
            // The analyzer process has finished, so we can resolve the promise
            // to allow the build process to exit gracefully.
            resolve();
        });
    });
}
